circuits/src/recipient.rs

```diff
@@ -0,0 +1,114 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_jubjub::JubJubScalar;
+use dusk_plonk::prelude::*;
+use jubjub_schnorr::{gadgets, Signature};
+
+use crate::elgamal;
+use phoenix_core::PublicKey;
+
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+
+use phoenix_core::SecretKey;
+
+const TX_OUTPUT_NOTES: usize = 2;
+
+/// Parameters needed to prove a recipient in-circuit
+#[derive(Debug, Clone, Copy)]
+pub struct RecipientParameters {
+    /// Public key of the transaction sender
+    pub sender_pk: PublicKey,
+    /// Note public keys of each note recipient
+    pub recipient_npk_vec: [JubJubAffine; TX_OUTPUT_NOTES],
+    /// Signatures of 'payload_hash' verifiable using 'pk_A' and 'pk_B'
+    pub sig_vec: [Signature; TX_OUTPUT_NOTES],
+    /// Asymmetric encryption of 'pk_A' using both recipients 'npk'
+    pub enc_A_vec: [(JubJubExtended, JubJubExtended); TX_OUTPUT_NOTES],
+    /// Asymmetric encryption of 'pk_B' using both recipients 'npk'
+    pub enc_B_vec: [(JubJubExtended, JubJubExtended); TX_OUTPUT_NOTES],
+    /// Randomness needed to encrypt/decrypt 'pk_A'
+    pub r_A_vec: [JubJubScalar; TX_OUTPUT_NOTES],
+    /// Randomness needed to encrypt/decrypt 'pk_B'
+    pub r_B_vec: [JubJubScalar; TX_OUTPUT_NOTES],
+}
+
+impl Default for RecipientParameters {
+    fn default() -> Self {
+        let mut rng = StdRng::seed_from_u64(0xbeef);
+
+        let sk = SecretKey::random(&mut rng);
+        let sender_pk = PublicKey::from(&sk);
+
+        Self {
+            sender_pk,
+            recipient_npk_vec: [
+                JubJubAffine::default(),
+                JubJubAffine::default(),
+            ],
+            sig_vec: [Signature::default(), Signature::default()],
+            enc_A_vec: [(JubJubExtended::default(), JubJubExtended::default());
+                TX_OUTPUT_NOTES],
+            enc_B_vec: [(JubJubExtended::default(), JubJubExtended::default());
+                TX_OUTPUT_NOTES],
+            r_A_vec: [JubJubScalar::default(); TX_OUTPUT_NOTES],
+            r_B_vec: [JubJubScalar::default(); TX_OUTPUT_NOTES],
+        }
+    }
+}
+
+/// Gadget to prove a valid origin for a given transaction.
+pub(crate) fn gadget(
+    composer: &mut Composer,
+    rp: &RecipientParameters,
+    payload_hash: &Witness,
+) -> Result<(), Error> {
+    // VERIFY A SIGNATURE FOR EACH KEY 'A' AND 'B'
+    let pk_A = composer.append_point(rp.sender_pk.A());
+    let pk_B = composer.append_point(rp.sender_pk.B());
+
+    let sig_A_u = composer.append_witness(*rp.sig_vec[0].u());
+    let sig_A_R = composer.append_point(rp.sig_vec[0].R());
+
+    let sig_B_u = composer.append_witness(*rp.sig_vec[1].u());
+    let sig_B_R = composer.append_point(rp.sig_vec[1].R());
+
+    gadgets::verify_signature(composer, sig_A_u, sig_A_R, pk_A, *payload_hash)?;
+    gadgets::verify_signature(composer, sig_B_u, sig_B_R, pk_B, *payload_hash)?;
```

Comment:
And you're dereferencing them as well, so it *really* should change.
```suggestion
    gadgets::verify_signature(composer, sig_A_u, sig_A_R, pk_A, payload_hash)?;
    gadgets::verify_signature(composer, sig_B_u, sig_B_R, pk_B, payload_hash)?;
```